enhancements/etcd/etcd-size-tuning.md

```diff
@@ -0,0 +1,226 @@
+
+## Proposal
+
+Allow the customer/admin access to set the backend quota directly through a human readable value via the etcd.operator.openshift.io/v1 techpreview field. The only restriction is that they are disallowed from setting the value lower than the current.
+
+Increasing this value will not have etcd immediately reserve this space on disk, it just allows etcd to grow up to that value if etcd grows under the usage of the cluster.
+
+The default value or unset value would be the current, hardcoded backend quota: 8GiB. This would allow for upgrades where this configuration value wouldn't nesscessarily be set.
+An initial upper limit of 64GiB will be enforced so as to limit the impact of some etcd operations, like restore, transporting snapshots, etc. 64GiB is an arbitrary value and work with the Perf-Scale team will need to be done to validate this value and potentially increase the upper limit; finding a reasonable and safe upper bound will need to be done before this feature is made GA.
+
+We will need to work with the perf team to understand the memory requirements of largers quotas, and the effect of the quota on things like defragmentation, snapshot restore, etc.
+A downside is that with the profiles, discussed in the Alternatives section, there are a discrete, small, number of permutations to test, giving more confidence of the exact effects a given profile has on performance/memory requirements.
+This is a minor consideration as setting the quota to a arbitrarily higher value does not mean that etcd will take up that much space; it will depend on the workload.
+
+### Workflow Description
+
+**cluster administrator** is a human user responsible deploying and maintaining the cluster.
+
+1. The cluster administrator decides to change the etcd backend quota from the default (8GiB) to a larger value; e.g. 20GiB.
+2. They set the new value via the `etcd.operator.openshift.io/v1` API.
+3. An etcd redeployment will be automatically run which restarts the etcd pods which consume the new value.
+
+#### Invalid value
+
+1. The cluster administrator wants to set the quota from a larger value (e.g. 32GiB) to a smaller one (8GiB).
+2. They attempt to set the new value via the `etcd.operator.openshift.io/v1` API.
+3. The API server should return an error notifying them that the value is invalid - not allowed to shrink the quota.
+
+### API Extensions
+
+In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
+- Add BackendQuotaGiB of type int64
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
+  - With an upper limit of '64GiB'.
+
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:default=8
+  // +kubebuilder:validation:Maximum=64
+  // +kubebuilder:validation:XValidation:rule="self.CmpInt64(oldSelf)>0",message="can't decrease database size"
```

Comment:
```suggestion
  // +kubebuilder:validation:XValidation:rule="self >= oldSelf",message="can't decrease database size"
```

Reply:
during running the e2e test against these changes, `CmpInt64` failed.

Reply:
int64 does not implement CmpInt64